src/ua_securechannel.c

```diff
@@ -230,189 +248,160 @@ void UA_SecureChannel_revolveTokens(UA_SecureChannel* channel) {
 
-    // Add signature length to message size
-    if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
-       channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        respHeader.messageHeader.messageSize += securityPolicy->symmetricModule.signingModule.signatureSize;
-    }
-
-    ci->messageSizeSoFar += respHeader.messageHeader.messageSize;
-
-    /* Encode the header at the beginning of the buffer */
     UA_SymmetricAlgorithmSecurityHeader symSecHeader;
     symSecHeader.tokenId = channel->securityToken.tokenId;
+    UA_encodeBinary(&symSecHeader.tokenId, &UA_TYPES[UA_TYPES_UINT32],
+                    &header_pos, buf_end, NULL, NULL);
+
     UA_SequenceHeader seqHeader;
     seqHeader.requestId = ci->requestId;
     seqHeader.sequenceNumber = UA_atomic_add(&channel->sendSequenceNumber, 1);
-    size_t offset_header = 0;
-    UA_SecureConversationMessageHeader_encodeBinary(&respHeader, dst, &offset_header);
-    UA_SymmetricAlgorithmSecurityHeader_encodeBinary(&symSecHeader, dst, &offset_header);
-    UA_SequenceHeader_encodeBinary(&seqHeader, dst, &offset_header);
+    UA_encodeBinary(&seqHeader, &UA_TRANSPORT[UA_TRANSPORT_SEQUENCEHEADER],
+                    &header_pos, buf_end, NULL, NULL);
 
-    // Sign message
+    /* Sign message */
     if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGN ||
        channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        const UA_ByteString dataToSign = {
-            offset,
-            dst->data
-        };
+        UA_ByteString dataToSign = ci->messageBuffer;
+        dataToSign.length = pre_sig_length;
 
-        UA_ByteString signature = {
-            securityPolicy->symmetricModule.signingModule.signatureSize,
-            dst->data + offset
-        };
+        UA_ByteString signature;
+        signature.length = securityPolicy->symmetricModule.signingModule.signatureSize;
+        signature.data = *buf_pos;
 
-        securityPolicy->symmetricModule.signingModule.sign(&dataToSign, channel->securityContext, &signature);
-
-        offset += signature.length;
+        securityPolicy->symmetricModule.signingModule.sign(&dataToSign, channel->securityContext,
+                                                           &signature);
     }
 
-    // Encrypt message
+    /* Encrypt message */
     if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        const UA_ByteString dataToEncrypt = {
-            offset - UA_SECUREMH_AND_SYMALGH_LENGTH,
-            dst->data + UA_SECUREMH_AND_SYMALGH_LENGTH
-        };
+        UA_ByteString dataToEncrypt;
+        dataToEncrypt.data = ci->messageBuffer.data + UA_SECUREMH_AND_SYMALGH_LENGTH;
+        dataToEncrypt.length = total_length - UA_SECUREMH_AND_SYMALGH_LENGTH;
 
+        /* TODO: Is it necessary to allocate data on the heap at this place?
+           Maybe we only need to allocate data for one block or hide this inside
+           the plugin callback. */
```

Comment:
> Though, does the standard specify if only encryption algs where the input length = output length are used?

I guess not. But I don't know a counter example. Maybe ask Christian.
Anyway, the message size computation would have to be revamped.
So I would not consider this for now.